cmd/skaffold/app/cmd/cmd.go

```diff
@@ -75,26 +82,29 @@ func SetUpLogs(out io.Writer, level string) error {
 	return nil
 }
 
-var defaultConfigForMode = map[bool]*config.SkaffoldConfig{
-	true:  config.DefaultDevSkaffoldConfig,
-	false: config.DefaultRunSkaffoldConfig,
-}
-
 func runSkaffold(out io.Writer, dev bool, filename string) error {
-	buf, err := readConfiguration(filename)
+	buf, err := util.ReadConfiguration(filename)
 	if err != nil {
 		return errors.Wrap(err, "read skaffold config")
 	}
 
-	defaultConfig := defaultConfigForMode[dev]
-	cfg, err := config.Parse(buf, defaultConfig)
+	apiVersion := &config.ApiVersion{}
+	if err := yaml.Unmarshal(buf, apiVersion); err != nil {
+		return errors.Wrap(err, "parsing api version")
+	}
+
+	cfg, err := config.GetConfig(buf, true, dev)
 	if err != nil {
 		return errors.Wrap(err, "parsing skaffold config")
 	}
 
+	if apiVersion.Version != config.LatestVersion {
+		return fmt.Errorf("Config version %s out of date with latest version %s: please run `skaffold fix`", apiVersion.Version, config.LatestVersion)
+	}
+
 	opts.Output = out
 	opts.DevMode = dev
-	r, err := runner.NewForConfig(opts, cfg)
+	r, err := runner.NewForConfig(opts, cfg.(*config.SkaffoldConfig))
```

Comment:
I don't particularly like this type assertion.

Reply:
I'm not the biggest fan of it either, but the alternative is exposing all fields in a config through getter methods and then refactoring all skaffold components to call those methods to retrieve the fields they need. this also binds us to those fields and makes it a lot harder to make major config changes down the road, since we'd need to go change all the callers of those methods. definitely open to suggestions here but I still think this is the best way to handle this